src/nix/nix_store.rs

```diff
@@ -127,11 +140,59 @@ impl NixStoreCmd {
                 .collect();
             Ok(out)
         } else {
-            let exit_code = out.status.code().unwrap_or(1);
-            bail!(
-                "nix-store --query --requisites --include-outputs failed to run (exited: {})",
-                exit_code
-            );
+            let stderr = String::from_utf8(out.stderr).ok();
+            let exit_code = out.status.code();
+            Err(CommandError::ProcessFailed { stderr, exit_code }.into())
         }
     }
 }
+
+/// Errors when running and interpreting the output of a nix command
+#[derive(Error, Debug)]
+pub enum NixCmdError {
+    #[error("Command error: {0}")]
+    CmdError(#[from] CommandError),
+
+    #[error("Failed to decode command stdout (utf8 error): {0}")]
+    DecodeErrorUtf8(#[from] std::string::FromUtf8Error),
+
+    #[error("Failed to decode command stdout (from_str error): {0}")]
+    DecodeErrorFromStr(#[from] FromStrError),
+
+    #[error("Failed to decode command stdout (json error): {0}")]
+    DecodeErrorJson(#[from] serde_json::Error),
+
+    #[error("Unknown deriver in drv_path")]
+    UnknownDeriverError,
+}
+
+#[derive(Debug)]
+pub struct FromStrError(String);
+
+impl Display for FromStrError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Failed to parse string: {}", self.0)
+    }
+}
+
+impl std::error::Error for FromStrError {}
+
+#[derive(Error, Debug)]
+pub enum CommandError {
+    #[error("Child process error: {0}")]
+    ChildProcessError(#[from] std::io::Error),
+    #[error(
+        "Process exited unsuccessfully. exit_code={:?}{}",
+        exit_code,
+        match stderr {
+            Some(s) => format!(" stderr={}", s),
+            None => "".to_string()
+        },
+    )]
+    ProcessFailed {
+        stderr: Option<String>,
+        exit_code: Option<i32>,
+    },
+    #[error("Failed to decode command stderr: {0}")]
+    Decode(#[from] std::string::FromUtf8Error),
+}
```

Comment:
Again; why copy this from nix-rs crate?

https://github.com/juspay/nix-rs/blob/2dbba678a9e67925d964dbde24fbd8951ecbd2de/src/command.rs#L213